onnxruntime/contrib_ops/cpu/qlinear_concat.cc

```diff
@@ -0,0 +1,139 @@
+  const Tensor* tensor_y_scale = ctx->Input<Tensor>(0);
+  const Tensor* tensor_y_zero_point = ctx->Input<Tensor>(1);
+  bool is_signed_int8 = tensor_y_zero_point->IsDataType<int8_t>();
+  const auto identity_float = [](float v) -> float { return v; };
+
+  // Number of input tensors to concatenate (tupled)
+  auto input_count_x3 = Node().InputArgCount()[2];
+  ORT_ENFORCE(input_count_x3 >= 6 && input_count_x3 % 3 == 0,
+              "At least two inputs are needed, and each input must be (tensor, scale, zero_point) tuple!");
+
+  // Hold pointers to the input tensors to be used in the PrepareForCompute() step
+  auto input_count = input_count_x3 / 3;
+  std::vector<std::vector<uint8_t>> dynamic_lookup_tables(input_count);
+
+  std::vector<const Tensor*> input_tensors(input_count);
+  for (auto input_index = 0; input_index < input_count; ++input_index) {
+    auto tuple_start = 2 + input_index * 3;
+    input_tensors[input_index] = ctx->Input<Tensor>(static_cast<int>(tuple_start));
+
+    // Build lookup table for non-const parameters (scale + zero point)
+    if (fixed_lookup_tables_[input_index].size() == 0) {
+      // Check tensor type first
+      const Tensor* tensor_x_scale = ctx->Input<Tensor>(tuple_start + 1);
+      const Tensor* tensor_x_zero_point = ctx->Input<Tensor>(tuple_start + 2);
+
+      ORT_ENFORCE(tensor_x_scale->IsDataType<float>(), "Input scale is not float for quantized input @", tuple_start + 1);
+      ORT_ENFORCE(tensor_x_zero_point->GetElementType() == tensor_y_zero_point->GetElementType(),
+                "Wrong input type encountered for zero point of quantized input @", tuple_start + 2);
+
+      dynamic_lookup_tables[input_index].resize(256);
+      if (is_signed_int8) {
+        QlinearBuildLookupTable<int8_t>(
+            dynamic_lookup_tables[input_index].data(), tensor_x_scale, tensor_x_zero_point,
+            tensor_y_scale, tensor_y_zero_point, identity_float);
+      } else {
+        QlinearBuildLookupTable<uint8_t>(
+            dynamic_lookup_tables[input_index].data(), tensor_x_scale, tensor_x_zero_point,
+            tensor_y_scale, tensor_y_zero_point, identity_float);
+      }
+    }
+  }
+
+  // Validate inputs and prepare some metadata used during actual compute
+  Prepare p;
+  auto status = PrepareForCompute(ctx, input_tensors, p);
+  if (!status.IsOK())
+    return status;
+
+  // Return at this point if output tensor is going to be empty
+  if (p.output_num_elements == 0)
+    return Status::OK();
+
+  int64_t initial_output_offset = 0;  // initial offset for each input
+  for (int input_index = 0; input_index < input_count; input_index++) {
+    const auto& prep = p.inputs[input_index];
+    if (prep.num_elements == 0)
+      continue;
+
+    const uint8_t* table = (fixed_lookup_tables_[input_index].size() > 0)
+                                ? fixed_lookup_tables_[input_index].data()
+                                : dynamic_lookup_tables[input_index].data();
```

Comment:
Can we build the dynamic lookup table for this input here, and remove the loop above, so we don't need that vector allocation?